src/Controllers/HistoryViewerController.php

```diff
@@ -339,4 +419,31 @@ protected function scaffoldForm($formName, array $context = [], array $extra = [
             LeftAndMainFormRequestHandler::create($form, $extra)
         );
     }
+
+    private function getPostedJsonValue(HTTPRequest $request, string $key): mixed
+    {
+        $data = json_decode($request->getBody(), true);
+        if (!array_key_exists($key, $data)) {
+            $this->jsonError(400);
+        }
+        return $data[$key];
+    }
+
+    private function getDataObject(int $id, string $recordClass, bool $use404): DataObject
+    {
+        if (!$id) {
+            $this->jsonError($use404 ? 404 : 400);
+        }
+        if (!$recordClass || !class_exists($recordClass) || !is_a($recordClass, DataObject::class, true)) {
+            $this->jsonError(400);
+        }
+        $obj = $recordClass::get()->byID($id);
+        if (!$obj) {
+            $this->jsonError($use404 ? 404 : 400);
+        }
+        if (!$obj->hasExtension(Versioned::class)) {
+            $this->jsonError(400);
+        }
+        return $obj;
+    }
```

Comment:
This feels very similar to the `getFileByID()` you added to `AssetAdminOpen` - should it be generalised and put somewhere central like `LeftAndMain`? That would reduce duplicate code and mean if we want to update it later we just need to do so in one place.

Reply:
It is, though it's currently only used by one class right now, and it's a private method so could easily be made more generic later

It's also specific to doing JSON requests, and it allows polymorphism, which we don't allow on any other controllers, so it seems pretty unlikely to be used anywhere else

Reply:
It would literally be used elsewhere, in `AssetAdminOpen` in place of `getFileByID()`. But like you say it's private so we can rip it out later when we look at these again and say "why is this almost identical logic in multiple places?"